src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+
+            try
+            {
+                while (true)
+                {
+                    var modifiers = _pool.Allocate<SyntaxToken>();
+                    try
+                    {
+                        // Modifiers will be checked for validity in later steps
+                        while (IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword)
+                        {
+                            modifiers.Add(EatToken());
+                        }
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.FunctionPointerParameterOrReturnType(modifiers, parameterType));
+
+                        if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        Debug.Assert(CurrentToken.Kind == SyntaxKind.CommaToken);
+                        types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                    }
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.CurrentToken.Kind == SyntaxKind.CommaToken || p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerStart()
+            => CurrentToken.Kind == SyntaxKind.DelegateKeyword && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
+
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
+               // Treat predefined types as possible calling conventions, as they're a single
+               // token. This will be a semantic error, but it'll result in a nicer parse tree.
+               (CurrentToken.Kind == SyntaxKind.IdentifierToken || IsPredefinedType(CurrentToken.Kind));
```

Comment:
why only predefined types?  why not any keyword?

Reply:
Again in search of nicer parse trees. I could understand putting a predefined type there accidentally because you thought that's where the return type went, but other keywords are almost certainly part of some other construct and it would be better to parse them as they are, rather than consuming that token and attempting to parse that thing. For example:
```cs
delegate* // Declaration in progress
while (true) { }
```
It would be better to just ignore the `while` here, even though the token that follows it _is_ the start of what we could consider a function parameter list start.

---
In reply to: [348850737](https://github.com/dotnet/roslyn/pull/39879#discussion_r348850737) [](ancestors = 348850737)

Reply:
another great reason :)   Can you please comment accordingly?  this sort of knowledge is great to have in the parser so people later on understand what's going on.